marconi/src/Marconi/Index/AddressDatum.hs

```diff
@@ -0,0 +1,517 @@
+
+    -- Datum hashes that are linked to an address, but do not have a corresponding datum value
+    -- associated with it.
+    let unresolvedDatumHashes =
+            Set.toList $ fold (Map.elems addressDatumMap) `Set.difference` Map.keysSet datumMap
+    datums <- forM unresolvedDatumHashes $ \dh -> do
+        (datum :: Maybe DatumRow) <- listToMaybe <$> SQL.query c
+          "SELECT datum_hash, datum FROM datumhash_datum WHERE datum_hash = ?" (SQL.Only dh)
+        pure $ fmap (\(DatumRow _ d) -> (dh, d)) datum
+    let resolvedDatumHashes = Map.fromList $ catMaybes datums
+
+    pure $ AddressDatumResult
+         $ storableEventToResult
+            q
+            (AddressDatumIndexEvent addressDatumMap (datumMap <> resolvedDatumHashes) cp)
+
+    where
+      storableEventToResult
+          :: C.AddressAny
+          -> StorableEvent AddressDatumHandle
+          -> Set C.ScriptData
+      storableEventToResult targetAddr (AddressDatumIndexEvent addressDatumMap datumMap _chainPoint) =
+          Set.map snd
+            $ Set.filter ((==) targetAddr . fst)
+            $ foldMap (\(addr, datumHashes) -> Set.map (addr,) $ resolveMapKeys datumHashes datumMap)
+            $ Map.toList addressDatumMap
+
+      resolveMapKeys
+          :: (Ord k, Ord v)
+          => Set k
+          -> Map k v
+          -> Set v
+      resolveMapKeys keys m =
+          -- TODO Not efficient to convert back n forth between Set
+          Set.fromList $ mapMaybe (\k -> Map.lookup k m) $ Set.toList keys
+
+instance Rewindable AddressDatumHandle where
+    rewindStorage
+        :: C.ChainPoint
+        -> AddressDatumHandle
+        -> IO (Maybe AddressDatumHandle )
+    rewindStorage C.ChainPointAtGenesis h@(AddressDatumHandle c _) = do
+         SQL.execute_ c "DELETE FROM address_datums"
+         pure $ Just h
+    rewindStorage (C.ChainPoint sn _) h@(AddressDatumHandle c _) = do
+         SQL.execute c "DELETE FROM address_datums WHERE slot_no > ?" (SQL.Only sn)
+         pure $ Just h
+
+instance Resumable AddressDatumHandle where
+    resumeFromStorage
+        :: AddressDatumHandle
+        -> IO [C.ChainPoint]
+    resumeFromStorage h = do
+        es <- Storable.getStoredEvents h
+        pure $ fmap (\(AddressDatumIndexEvent _ _ chainPoint) -> chainPoint) es
+            ++ [C.ChainPointAtGenesis]
+
+open
+  :: FilePath
+  -> AddressDatumDepth
+  -> IO AddressDatumIndex
+open dbPath (AddressDatumDepth k) = do
+    c <- SQL.open dbPath
+
+    SQL.execute_ c
+        [r|CREATE TABLE IF NOT EXISTS address_datums
+            ( address TEXT NOT NULL
+            , datum_hash BLOB NOT NULL
+            , slot_no INT NOT NULL
+            , block_hash BLOB NOT NULL
+            )|]
+    SQL.execute_ c
+        [r|CREATE TABLE IF NOT EXISTS datumhash_datum
+            ( datum_hash BLOB PRIMARY KEY
+            , datum BLOB
+            )|]
+    SQL.execute_ c
```

Comment:
Is it possible to have duplicate entries? should we add uniqueness constraint?

Reply:
Is it possible to have multiple identical rows, and is it OK to have them?

Reply:
Are you referring to the index table `address_datums_index`? If so, yes, one address can have multiple datums.

Reply:
right. my question however is with this  table structure, it is possible to have duplicate rows. Is that OK? and consequently, when we query, and get a `list` back, is it OK for the results set to have identical items?

Reply:
I'm not seeing it. How can a query to `address_datums_index` result in duplicate rows? 
Querying an address will yield a list of datum hashes. Of course, there should no be any duplicate datum hashes for a given address. Should I change anything for that?

Reply:
thnx, @koslambrou that was mostly my question that is it possible to have duplicate hashes, and the answer is no.